docs/architecture/adr-033-protobuf-inter-module-rpc.md

```diff
@@ -0,0 +1,323 @@
+
+func (am AppModule) RegisterServices(cfg Configurator) {
+  cfg.RequireServer((*bank.QueryServer)(nil))
+  cfg.RequireServer((*bank.MsgServer)(nil))
+}
+```
+
+### Security Considerations
+
+In addition to checking for `ModuleKey` permissions, a few additional security precautions will need to be taken by
+the underlying router infrastructure.
+
+#### Recursion and Re-entry
+
+Recursive or re-entrant method invocations pose a potential security threat. This can be a problem if Module A
+calls Module B and Module B calls module A again in the same call.
+
+One basic way for the router system to deal with this is to maintain a call stack which prevents a module from
+being referenced more than once in the call stack so that there is no re-entry. A `map[string]interface{}` table
+in the router could be used to perform this security check.
+
+#### Queries
+
+Queries in Cosmos SDK are generally un-permissioned so allowing one module to query another module should not pose
+any major security threats assuming basic precautions are taken. The basic precaution that the router system will
+need to take is making sure that the `sdk.Context` passed to query methods does not allow writing to the store. This
+can be done for now with a `CacheMultiStore` as is currently done for `BaseApp` queries.
+
+### Internal Methods
+
+In many cases, we may wish for modules to call methods on other modules which are not exposed to clients at all. For this
+purpose, we add the `InternalServer` method to `Configurator`:
+
+```go
+type Configurator interface {
+   MsgServer() grpc.Server
+   QueryServer() grpc.Server
+   InternalServer() grpc.Server
+}
+```
+
+Services registered against `InternalServer` will be callable from other modules but not by external clients.
+
+### Authorization
+
+By default, the inter-module router requires that modules are sent by the first signer returned by `GetSigners`. The
+inter-module router should also accept authorization middleware such as that provided by [ADR 030](https://github.com/cosmos/cosmos-sdk/pull/7105).
+This middleware will allow accounts to otherwise specific module accounts to perform actions on their behalf.
+Authorization middleware should take into account the need to grant certain modules effectively "admin" privileges to
+other modules. This will be addressed in separate ADRs or updates to this ADR.
+
+### Future Work
+
+Other future improvements may include:
+* custom code generation that:
+  * simplifies interfaces (ex. generates code with `sdk.Context` instead of `context.Context`)
+  * optimizes inter-module calls - for instance caching resolved methods after first invocation
+* combining `StoreKey`s and `ModuleKey`s into a single interface so that modules have a single Ocaps handle
+* code generation which makes inter-module communication more performant
+* decoupling `ModuleKey` creation from `AppModuleBasic.Name()` so that app's can override root module account names
+* inter-module hooks and plugins
```

Comment:
Also added a section on custom code generation and after adding `InternalServer` am still thinking to do hooks/plugins separately (still working on design details). /cc @amaurymartiny